client/genericclient/stream.go

```diff
@@ -0,0 +1,220 @@
+
+	kc, err := client.NewClient(svcInfo, options...)
+	if err != nil {
+		return nil, err
+	}
+	cli := &genericServiceClient{
+		svcInfo: svcInfo,
+		kClient: kc,
+		g:       g,
+	}
+	runtime.SetFinalizer(cli, (*genericServiceClient).Close)
+
+	svcInfo.GenericMethod = func(name string) serviceinfo.MethodInfo {
+		n, err := g.GetMethod(nil, name)
+		var key string
+		if err != nil {
+			// TODO: support fallback solution for binary
+			key = serviceinfo.GenericMethod
+		} else {
+			key = getGenericStreamingMethodInfoKey(n.StreamingMode)
+		}
+		m := svcInfo.Methods[key]
+		return &methodInfo{
+			MethodInfo: m,
+			oneway:     n.Oneway,
+		}
+	}
+
+	return cli, nil
+}
+
+func getGenericStreamingMethodInfoKey(streamingMode serviceinfo.StreamingMode) string {
+	switch streamingMode {
+	case serviceinfo.StreamingClient:
+		return serviceinfo.GenericClientStreamingMethod
+	case serviceinfo.StreamingServer:
+		return serviceinfo.GenericServerStreamingMethod
+	case serviceinfo.StreamingBidirectional:
+		return serviceinfo.GenericBidirectionalStreamingMethod
+	default:
+		return serviceinfo.GenericMethod
+	}
+}
+
+type clientStreamingClient struct {
+	streaming.Stream
+	svcInfo *serviceinfo.ServiceInfo
+	method  string
+}
+
+func NewClientStreaming(ctx context.Context, genericCli Client, method string, callOpts ...callopt.Option) (ClientStreaming, error) {
+	gCli, ok := genericCli.(*genericServiceClient)
+	if !ok {
+		return nil, errors.New("invalid generic client")
+	}
+	stream, err := getStream(ctx, genericCli, method, callOpts...)
+	if err != nil {
+		return nil, err
+	}
+	return &clientStreamingClient{stream, gCli.svcInfo, method}, nil
+}
+
+func (cs *clientStreamingClient) Send(req interface{}) error {
+	_args := cs.svcInfo.MethodInfo(cs.method).NewArgs().(*generic.Args)
+	_args.Method = cs.method
+	_args.Request = req
+	return cs.Stream.SendMsg(_args)
+}
+
+func (cs *clientStreamingClient) CloseAndRecv() (resp interface{}, err error) {
+	if err := cs.Stream.Close(); err != nil {
+		return nil, err
+	}
+	_result := cs.svcInfo.MethodInfo(cs.method).NewResult().(*generic.Result)
+	if err = cs.Stream.RecvMsg(_result); err != nil {
+		return nil, err
+	}
+	return _result.GetSuccess(), nil
+}
+
+type serverStreamingClient struct {
+	streaming.Stream
+	svcInfo *serviceinfo.ServiceInfo
+	method  string
+}
+
+func NewServerStreaming(ctx context.Context, genericCli Client, method string, req interface{}, callOpts ...callopt.Option) (ServerStreaming, error) {
+	gCli, ok := genericCli.(*genericServiceClient)
+	if !ok {
+		return nil, errors.New("invalid generic client")
+	}
+	stream, err := getStream(ctx, genericCli, method, callOpts...)
+	if err != nil {
+		return nil, err
+	}
+	ss := &serverStreamingClient{Stream: stream, svcInfo: gCli.svcInfo, method: method}
+	_args := gCli.svcInfo.MethodInfo(method).NewArgs().(*generic.Args)
+	_args.Method = method
+	_args.Request = req
+	if err = ss.Stream.SendMsg(_args); err != nil {
```

Comment:
why we need to `Send()` and `Close()` first?

Reply:
@AsterDY I think this is what server streaming client is meant to do. client doesn't send streaming request (= send only one request), so we can send the request and close the stream here.

Reply:
@felix021 Do you know why?

Reply:
It's by design: 
- For a streaming client, `Close` actually means `CloseSend`
- It sends a End-Of-Stream message (in GRPC, a frame with the flag set) to the server, so the `recv` call in server side can return an `io.EOF`
- A ServerStreaming client should send exactly 1 message to the server, and calling `Close` ensures this.

Reply:
So `ss.Stream.SendMsg(_args)` is not neccessary? and What if we don't `CloseSend`?

Reply:
- `SendMsg(_args)` is necessary here to send the request.
- If `CloseSend` is not called, it might result in the server waiting forever on the 2nd recv call (There's no way to ensure the server calls recv only once on a server streaming method request).